indexer/services/vulcan/src/lib/on-batch.ts

```diff
@@ -0,0 +1,92 @@
+import { logger, stats } from '@dydxprotocol-indexer/base';
+import {
+  Batch,
+  EachBatchPayload,
+  KafkaMessage,
+} from 'kafkajs';
+
+import config from '../config';
+import { onMessage } from './on-message';
+
+export async function onBatch(
+  payload: EachBatchPayload,
+): Promise<void> {
+  const batch: Batch = payload.batch;
+  const topic: string = batch.topic;
+  const partition: string = batch.partition.toString();
+  const metricTags: Record<string, string> = { topic, partition };
+  if (batch.isEmpty()) {
+    logger.error({
+      at: 'on-batch#onBatch',
+      message: 'Empty batch',
+      ...metricTags,
+    });
+    return;
+  }
+
+  const startTime: number = Date.now();
+  const firstMessageTimestamp: number = Number(batch.messages[0].timestamp);
+  const batchTimeInQueue: number = startTime - firstMessageTimestamp;
+  const batchInfo = {
+    firstMessageTimestamp: new Date(firstMessageTimestamp).toISOString(),
+    batchTimeInQueue,
+    messagesInBatch: batch.messages.length,
+    firstOffset: batch.firstOffset(),
+    lastOffset: batch.lastOffset(),
+    ...metricTags,
+  };
+
+  logger.info({
+    at: 'on-batch#onBatch',
+    message: 'Received batch',
+    ...batchInfo,
+  });
+  stats.timing(
+    'vulcan.batch_time_in_queue',
+    batchTimeInQueue,
+    metricTags,
+  );
+
+  let lastCommitTime: number = startTime;
+  for (let i = 0; i < batch.messages.length; i++) {
+    const message: KafkaMessage = batch.messages[i];
+    await onMessage(message);
```

Comment:
One thing we did in v3 I think is use a [worker pool ](https://github.com/dydxprotocol/archivist/blob/26de6a4e4f1631b36a0912f2fe8296d96ce2537d/src/lib/on-batch.ts#L67)to process in parallel, but then to do it in chunks so that we could commit to an offset knowing that we had processed up to a current offset

Would we want to use the same strategy here?

Reply:
I don't think it works the same. The reason I designed the v3 solution to work in chunks was because archivist could handle messages out of order, but vulcan needs all messages to be processed in order, so there cannot be parallelization between messages of the same partition without additional knowledge of the message